# Project Team ID: PTID-CDS-DEC-24-2218

## Project Name: PRCP-1003-Customer Transaction Prediction

### Problem Statement:

#### Task 1:-Prepare a complete data analysis report on the given data.
#### Task 2:-Create a predictive model which will help the bank to identify which customer will make transactions in future.

### Aim of the Project:

#### --> Identify which customers will make a specific transaction in the future, irrespective of the amount transacted..

### Domain Expertise:

#### The banking domain refers to the industry and processes involved in managing financial assets, services, and systems. It encompasses a variety of sectors, including retail banking, corporate banking, investment banking, and central banking. The focus in this domain is on managing customers, financial products, transactions, risk, and compliance, leveraging technology and analytics.
#### 1. Retail Banking
##### --> Focuses on individual customers.
##### --> Services include savings accounts, loans, mortgages, credit cards, and wealth management.
#### 2. Corporate Banking
##### -->Caters to businesses, corporations, and institutions.
##### -->Services include business loans, trade finance, and treasury services.
#### 3. Investment Banking
##### --> Involves capital raising, mergers and acquisitions, and trading financial instruments.
#### 4. Central Banking
##### -->Manages monetary policy and regulates commercial banks.
##### -->Ensures financial stability and liquidity.

# Task 1:

### Importing Necessary Libraries

In [1]:
import numpy as np
import pandas as pd                # importing the necessary Libraries 
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
data=pd.read_csv("https://d3ilbtxij3aepc.cloudfront.net/projects/CDS-Capstone-Projects/PRCP-1003-CustTransPred.zip")
data                               # loading the data

ID_code  target    var_0   var_1    var_2   var_3    var_4  \
0            train_0       0   8.9255 -6.7863  11.9081  5.0930  11.4607   
1            train_1       0  11.5006 -4.1473  13.8588  5.3890  12.3622   
2            train_2       0   8.6093 -2.7457  12.0805  7.8928  10.5825   
3            train_3       0  11.0604 -2.1518   8.9522  7.1957  12.5846   
4            train_4       0   9.8369 -1.4834  12.8746  6.6375  12.2772   
...              ...     ...      ...     ...      ...     ...      ...   
199995  train_199995       0  11.4880 -0.4956   8.2622  3.5142  10.3404   
199996  train_199996       0   4.9149 -2.4484  16.7052  6.6345   8.3096   
199997  train_199997       0  11.2232 -5.0518  10.5127  5.6456   9.3410   
199998  train_199998       0   9.7148 -8.6098  13.6104  5.7930  12.5173   
199999  train_199999       0  10.8762 -5.7105  12.1183  8.0328  11.5577   

          var_5   var_6    var_7  ...  var_190  var_191  var_192  var_193  \
0       -9.2834  5.1187  18.6266  ...   4.4354   3.9642   3.1364   1.6910   
1        7.0433  5.6208  16.5338  ...   7.6421   7.7214   2.5837  10.9516   
2       -9.0837  6.9427  14.6155  ...   2.9057   9.7905   1.6704   1.6858   
3       -1.8361  5.8428  14.9250  ...   4.4666   4.7433   0.7178   1.4214   
4        2.4486  5.9405  19.2514  ...  -1.4905   9.5214  -0.1508   9.1942   
...         ...     ...      ...  ...      ...      ...      ...      ...   
199995  11.6081  5.6709  15.1516  ...   6.1415  13.2305   3.9901   0.9388   
199996 -10.5628  5.8802  21.5940  ...   4.9611   4.6549   0.6998   1.8341   
199997  -5.4086  4.5555  21.5571  ...   4.0651   5.4414   3.1032   4.8793   
199998   0.5339  6.0479  17.0152  ...   2.6840   8.6587   2.7337  11.1178   
199999   0.3488  5.2839  15.2058  ...   8.9842   1.6893   0.1276   0.3766   

        var_194  var_195  var_196  var_197  var_198  var_199  
0       18.5227  -2.3978   7.8784   8.5635  12.7803  -1.0914  
1       15.4305   2.0339   8.1267   8.7889  18.3560   1.9518  
2       21.6042   3.1417  -6.5213   8.2675  14.7222   0.3965  
3       23.0347  -1.2706  -2.9275  10.2922  17.9697  -8.9996  
4       13.2876  -1.5121   3.9267   9.5031  17.9974  -8.8104  
...         ...      ...      ...      ...      ...      ...  
199995  18.0249  -1.7939   2.1661   8.5326  16.6660 -17.8661  
199996  22.2717   1.7337  -2.1651   6.7419  15.9054   0.3388  
199997  23.5311  -1.5736   1.2832   8.7155  13.8329   4.1995  
199998  20.4158  -0.0786   6.7980  10.0342  15.5289 -13.9001  
199999  15.2101  -2.4907  -2.2342   8.1857  12.1284   0.1385  

[200000 rows x 202 columns]

In [3]:
data.head()                        # shows the first 5 rows and all columns from the dataset

ID_code  target    var_0   var_1    var_2   var_3    var_4   var_5   var_6  \
0  train_0       0   8.9255 -6.7863  11.9081  5.0930  11.4607 -9.2834  5.1187   
1  train_1       0  11.5006 -4.1473  13.8588  5.3890  12.3622  7.0433  5.6208   
2  train_2       0   8.6093 -2.7457  12.0805  7.8928  10.5825 -9.0837  6.9427   
3  train_3       0  11.0604 -2.1518   8.9522  7.1957  12.5846 -1.8361  5.8428   
4  train_4       0   9.8369 -1.4834  12.8746  6.6375  12.2772  2.4486  5.9405   

     var_7  ...  var_190  var_191  var_192  var_193  var_194  var_195  \
0  18.6266  ...   4.4354   3.9642   3.1364   1.6910  18.5227  -2.3978   
1  16.5338  ...   7.6421   7.7214   2.5837  10.9516  15.4305   2.0339   
2  14.6155  ...   2.9057   9.7905   1.6704   1.6858  21.6042   3.1417   
3  14.9250  ...   4.4666   4.7433   0.7178   1.4214  23.0347  -1.2706   
4  19.2514  ...  -1.4905   9.5214  -0.1508   9.1942  13.2876  -1.5121   

   var_196  var_197  var_198  var_199  
0   7.8784   8.5635  12.7803  -1.0914  
1   8.1267   8.7889  18.3560   1.9518  
2  -6.5213   8.2675  14.7222   0.3965  
3  -2.9275  10.2922  17.9697  -8.9996  
4   3.9267   9.5031  17.9974  -8.8104  

[5 rows x 202 columns]

In [4]:
data.tail()                        # shows the last 5 rows and all columns from the dataset

ID_code  target    var_0   var_1    var_2   var_3    var_4  \
199995  train_199995       0  11.4880 -0.4956   8.2622  3.5142  10.3404   
199996  train_199996       0   4.9149 -2.4484  16.7052  6.6345   8.3096   
199997  train_199997       0  11.2232 -5.0518  10.5127  5.6456   9.3410   
199998  train_199998       0   9.7148 -8.6098  13.6104  5.7930  12.5173   
199999  train_199999       0  10.8762 -5.7105  12.1183  8.0328  11.5577   

          var_5   var_6    var_7  ...  var_190  var_191  var_192  var_193  \
199995  11.6081  5.6709  15.1516  ...   6.1415  13.2305   3.9901   0.9388   
199996 -10.5628  5.8802  21.5940  ...   4.9611   4.6549   0.6998   1.8341   
199997  -5.4086  4.5555  21.5571  ...   4.0651   5.4414   3.1032   4.8793   
199998   0.5339  6.0479  17.0152  ...   2.6840   8.6587   2.7337  11.1178   
199999   0.3488  5.2839  15.2058  ...   8.9842   1.6893   0.1276   0.3766   

        var_194  var_195  var_196  var_197  var_198  var_199  
199995  18.0249  -1.7939   2.1661   8.5326  16.6660 -17.8661  
199996  22.2717   1.7337  -2.1651   6.7419  15.9054   0.3388  
199997  23.5311  -1.5736   1.2832   8.7155  13.8329   4.1995  
199998  20.4158  -0.0786   6.7980  10.0342  15.5289 -13.9001  
199999  15.2101  -2.4907  -2.2342   8.1857  12.1284   0.1385  

[5 rows x 202 columns]

In [5]:
data.describe()                    # Statistical Analytics i.e count, mean ,standard, min,25%, 50% 75% and max

target          var_0          var_1          var_2  \
count  200000.000000  200000.000000  200000.000000  200000.000000   
mean        0.100490      10.679914      -1.627622      10.715192   
std         0.300653       3.040051       4.050044       2.640894   
min         0.000000       0.408400     -15.043400       2.117100   
25%         0.000000       8.453850      -4.740025       8.722475   
50%         0.000000      10.524750      -1.608050      10.580000   
75%         0.000000      12.758200       1.358625      12.516700   
max         1.000000      20.315000      10.376800      19.353000   

               var_3          var_4          var_5          var_6  \
count  200000.000000  200000.000000  200000.000000  200000.000000   
mean        6.796529      11.078333      -5.065317       5.408949   
std         2.043319       1.623150       7.863267       0.866607   
min        -0.040200       5.074800     -32.562600       2.347300   
25%         5.254075       9.883175     -11.200350       4.767700   
50%         6.825000      11.108250      -4.833150       5.385100   
75%         8.324100      12.261125       0.924800       6.003000   
max        13.188300      16.671400      17.251600       8.447700   

               var_7          var_8  ...        var_190        var_191  \
count  200000.000000  200000.000000  ...  200000.000000  200000.000000   
mean       16.545850       0.284162  ...       3.234440       7.438408   
std         3.418076       3.332634  ...       4.559922       3.023272   
min         5.349700     -10.505500  ...     -14.093300      -2.691700   
25%        13.943800      -2.317800  ...      -0.058825       5.157400   
50%        16.456800       0.393700  ...       3.203600       7.347750   
75%        19.102900       2.937900  ...       6.406200       9.512525   
max        27.691800      10.151300  ...      18.440900      16.716500   

             var_192        var_193        var_194        var_195  \
count  200000.000000  200000.000000  200000.000000  200000.000000   
mean        1.927839       3.331774      17.993784      -0.142088   
std         1.478423       3.992030       3.135162       1.429372   
min        -3.814500     -11.783400       8.694400      -5.261000   
25%         0.889775       0.584600      15.629800      -1.170700   
50%         1.901300       3.396350      17.957950      -0.172700   
75%         2.949500       6.205800      20.396525       0.829600   
max         8.402400      18.281800      27.928800       4.272900   

             var_196        var_197        var_198        var_199  
count  200000.000000  200000.000000  200000.000000  200000.000000  
mean        2.303335       8.908158      15.870720      -3.326537  
std         5.454369       0.921625       3.010945      10.438015  
min       -14.209600       5.960600       6.299300     -38.852800  
25%        -1.946925       8.252800      13.829700     -11.208475  
50%         2.408900       8.888200      15.934050      -2.819550  
75%         6.556725       9.593300      18.064725       4.836800  
max        18.321500      12.000400      26.079100      28.500700  

[8 rows x 201 columns]

In [6]:
data.info()                        #it shows the infomation of index ,no.of columns, datatypes and the memory usage

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200000 entries, 0 to 199999
Columns: 202 entries, ID_code to var_199
dtypes: float64(200), int64(1), object(1)
memory usage: 308.2+ MB


# Task 2:

### Feature Selection

In [7]:
X = data.drop(['ID_code','target'], axis=1)
y = data['target']                 # drop the ID_code and target column so that we dont want that columns to train the modal

In [8]:
y                                  # taking the target data a target variable

0         0
1         0
2         0
3         0
4         0
         ..
199995    0
199996    0
199997    0
199998    0
199999    0
Name: target, Length: 200000, dtype: int64

### Model Creation

In [11]:
## Step 1 Creating training and testing data.
from sklearn.model_selection import train_test_split
X_train, X_test, y_train , y_test = train_test_split(X,y,test_size=0.30,random_state=2)

In [12]:
y_test

105775    1
13154     0
181761    0
191939    0
149011    0
         ..
79398     0
118616    0
148225    0
36310     0
126529    0
Name: target, Length: 60000, dtype: int64

In [13]:
## Step 2 Model creation
from sklearn.linear_model import LogisticRegression
clf=LogisticRegression()
clf.fit(X_train,y_train)  ## training

C:\Users\Dell\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [14]:
## Step 3 Prediction
y_pred=clf.predict(X_test)            # predicting the y_predict by using x_test

In [15]:
y_pred                                

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

In [16]:
y_pred_prob=clf.predict_proba(X_test)   # probability of y_predict by using x_test

In [17]:
y_pred_prob

array([[0.97540182, 0.02459818],
       [0.99764199, 0.00235801],
       [0.95802284, 0.04197716],
       ...,
       [0.92291571, 0.07708429],
       [0.98675364, 0.01324636],
       [0.93231397, 0.06768603]])

In [18]:
data.target.value_counts()              # no. of transactions for the present data

target
0    179902
1     20098
Name: count, dtype: int64

### Model Evalution

In [19]:
from sklearn.metrics import confusion_matrix,accuracy_score,recall_score,precision_score,classification_report,f1_score

In [20]:
cm=confusion_matrix(y_test,y_pred)
print(cm)                                 # to evaluate the performance of a classification model

[[53185   649]
 [ 4779  1387]]


In [23]:
recall=recall_score(y_test,y_pred)
recall                                    # metric that measures how well a model identifies positive cases

0.22494323710671424

In [21]:
precision=precision_score(y_test,y_pred)
precision                                 # the quality of a positive prediction made by the model

0.681237721021611

In [22]:
f1score=f1_score(y_test,y_pred)
f1score                                   # that measures a model's accuracy by combining its precision and recall scores

0.33821019263594243

In [24]:
cr=classification_report(y_test,y_pred)
print(cr)                                  # a text summary that shows the main metrics for each class of a machine learning model

              precision    recall  f1-score   support

           0       0.92      0.99      0.95     53834
           1       0.68      0.22      0.34      6166

    accuracy                           0.91     60000
   macro avg       0.80      0.61      0.64     60000
weighted avg       0.89      0.91      0.89     60000



In [25]:
y_test.value_counts()                # no.of tranasctions happens and the no.of non tranasctions happens in future

target
0    53834
1     6166
Name: count, dtype: int64

# Conclusion

--> Target = 0: 53834 customers are predicted not to make a transaction.
    Target = 1: 6166 customers are predicted to make a transaction.

--> Most customers are unlikely to make transactions (89.72% are 0 and 10.27% are 1).
                                                  
--> The dataset is imbalanced, with far fewer customers predicted to make transactions(1)